File: Experimental/2023/Drivebase Test + Photon Vision/src/main/cpp/TrajectoryGenerator.cpp
#include "TrajectoryGenerator.h"

#include <frc/controller/SimpleMotorFeedforward.h>
#include <frc/trajectory/constraint/DifferentialDriveVoltageConstraint.h>
#include <frc/trajectory/TrajectoryConfig.h>
#include <frc/trajectory/TrajectoryGenerator.h>
#include <frc/geometry/Pose2d.h>
#include <frc/controller/RamseteController.h>
#include <frc/controller/PIDController.h>
#include <frc2/command/RamseteCommand.h>
#include <frc2/command/Commands.h>
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc2/command/InstantCommand.h>

#include <frc/Filesystem.h>
#include <frc/trajectory/TrajectoryUtil.h>
#include <wpi/fs.h>
#include <iostream>

#include "Constants.h"
#include "subsystems/Drivebase.h"

frc2::CommandPtr GetCommandForTrajectory(std::string fileToLoad, Drivebase* driveBase) {
// Create a voltage constraint to ensure we don't accelerate too fast
  frc::DifferentialDriveVoltageConstraint autoVoltageConstraint{
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics, 10_V};

  // Set up config for trajectory
  frc::TrajectoryConfig config{kMaxSpeed, kMaxAcceleration};
  // Add kinematics to ensure max speed is actually obeyed
  config.SetKinematics(kDriveKinematics);
  // Apply the voltage constraint
  config.AddConstraint(autoVoltageConstraint);

  config.SetReversed(false);

  // An example trajectory to follow.  All units in meters.
  
  /*auto exampleTrajectory = frc::TrajectoryGenerator::GenerateTrajectory(
      // Start at the origin facing the +X direction
      frc::Pose2d{0_m, 0_m, 0_deg},
      // Pass through these two interior waypoints, making an 's' curve path
      {
        frc::Translation2d{0_m, 0_m}
        //frc::Translation2d{-1_m, 1_m}
      },
      // End 3 meters straight ahead of where we started, facing forward
      frc::Pose2d{1_m, 0_m, 0_deg},
      // Pass the config
      config);*/

   fs::path deployDirectory = frc::filesystem::GetDeployDirectory();
   deployDirectory = deployDirectory / "paths" / fileToLoad.c_str();
   //Comment this back in and comment out the thing above to get the file
   frc::Trajectory exampleTrajectory  = frc::TrajectoryUtil::FromPathweaverJson(deployDirectory.string());

  frc2::CommandPtr ramseteCommand{frc2::RamseteCommand(
      exampleTrajectory, [driveBase] { return driveBase->GetPose(); }, // Displaying Get Pose in Dashboard
      frc::RamseteController{kRamseteB, kRamseteZeta},
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics,
      [driveBase] { return driveBase->GetWheelSpeeds(); },
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      [driveBase](auto left, auto right) { driveBase->TankDriveVolts(left, right); },
      {driveBase})}; 

  // Reset odometry to the starting pose of the trajectory.
  driveBase->ResetOdometry(exampleTrajectory.InitialPose());

  return std::move(ramseteCommand)
      .BeforeStarting(
          frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}))
      // Because Mr. Healy is professionally paranoid....
      .AndThen(frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}));
}




frc2::CommandPtr BuildTrajectoryUsingAprilTags(frc::Pose2d startingPose, frc::Pose2d endingPose, bool driveForward, Drivebase* driveBase){
    // Create a voltage constraint to ensure we don't accelerate too fast
  frc::DifferentialDriveVoltageConstraint autoVoltageConstraint{
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics, 10_V};

  // Set up config for trajectory
  frc::TrajectoryConfig config{kMaxSpeed, kMaxAcceleration};
  // Add kinematics to ensure max speed is actually obeyed
  config.SetKinematics(kDriveKinematics);
  // Apply the voltage constraint
  config.AddConstraint(autoVoltageConstraint);

  config.SetReversed(!driveForward);

  // An example trajectory to follow.  All units in meters.
  
  auto exampleTrajectory = frc::TrajectoryGenerator::GenerateTrajectory(
      startingPose,
      {},
      endingPose,
      config);


    for (const auto& state : exampleTrajectory.States()) {
    std::cout << "  "
       << "t: " << state.t.value() << ", vel: " << state.velocity.value()
       << ", a: " << state.acceleration.value() << ", pose: ("
       << state.pose.X().value() << "," << state.pose.Y().value() << ")"
       << "\n";
  }
  std::cout << "}\n";
//IMPORTANT CHANGE******** CHANGED driveBase->GetPose() to driveBase->GetEstimatedPose()
  frc2::CommandPtr ramseteCommand{frc2::RamseteCommand(
      exampleTrajectory, [driveBase] { return driveBase->GetEstimatedPose(); }, // Displaying Get Pose in Dashboard
      frc::RamseteController{kRamseteB, kRamseteZeta},
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics,
      [driveBase] { return driveBase->GetWheelSpeeds(); },
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      [driveBase](auto left, auto right) { driveBase->TankDriveVolts(left, right); },
      {driveBase})}; 

  // Reset odometry to the starting pose of the trajectory.
  driveBase->ResetOdometry(exampleTrajectory.InitialPose()); 

  return std::move(ramseteCommand)
      .BeforeStarting(
          frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}))
      // Because Mr. Healy is professionally paranoid....
      .AndThen(frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}));
}


/*void PerliminarySetup(){

};*/



//PLEASE REMMEMBER WHEN BUILDING MANUALLY THE ROBOT RESETS IT PLOTS THE COORDINATE PLANE ON TOP OF ITSELF SO WH
//WHATEVER DIRECTION IT IS FACING BECOMES THE X DIRECTION
frc2::CommandPtr ManuallyBuildTrajectoryUsingStandardOdometry(frc::Pose2d startingPose, std::vector<frc::Translation2d>& interiorWaypoints, frc::Pose2d endingPose, bool driveForward, Drivebase* driveBase)
{
    
      frc::DifferentialDriveVoltageConstraint autoVoltageConstraint{
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics, 10_V};

  // Set up config for trajectory
  frc::TrajectoryConfig config{kMaxSpeed, kMaxAcceleration};
  // Add kinematics to ensure max speed is actually obeyed
  config.SetKinematics(kDriveKinematics);
  // Apply the voltage constraint
  config.AddConstraint(autoVoltageConstraint);

  config.SetReversed(!driveForward);

  // An example trajectory to follow.  All units in meters.
  
  auto exampleTrajectory = frc::TrajectoryGenerator::GenerateTrajectory(
      // Start at the origin facing the +X direction
      startingPose,
      // Pass through these two interior waypoints, making an 's' curve path
      interiorWaypoints,
      // End 3 meters straight ahead of where we started, facing forward
      endingPose,
      // Pass the config
      config);

  frc2::CommandPtr ramseteCommand{frc2::RamseteCommand(
      exampleTrajectory, [driveBase] { return driveBase->GetPose(); }, // Displaying Get Pose in Dashboard
      frc::RamseteController{kRamseteB, kRamseteZeta},
      frc::SimpleMotorFeedforward<units::meters>{
          PathWeaverConstants::kS, PathWeaverConstants::kV, PathWeaverConstants::kA},
      kDriveKinematics,
      [driveBase] { return driveBase->GetWheelSpeeds(); },
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      frc2::PIDController{PathWeaverConstants::kP, 0, 0},
      [driveBase](auto left, auto right) { driveBase->TankDriveVolts(left, right); },
      {driveBase})}; 

  // Reset odometry to the starting pose of the trajectory.
  driveBase->ResetOdometry(exampleTrajectory.InitialPose());

  return std::move(ramseteCommand)
      .BeforeStarting(
          frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}))
      // Because Mr. Healy is professionally paranoid....
      .AndThen(frc2::cmd::RunOnce([driveBase] { driveBase->TankDriveVolts(0_V, 0_V); }, {}));
}

File: Quasics2026Code/CrescendoJava-2026/src/main/java/frc/robot/Trajectorygenerator.java
package frc.robot;

import static edu.wpi.first.units.Units.*;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.RamseteController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
import edu.wpi.first.math.trajectory.Trajectory;
import edu.wpi.first.math.trajectory.Trajectory.State;
import edu.wpi.first.math.trajectory.TrajectoryConfig;
import edu.wpi.first.math.trajectory.TrajectoryUtil;
import edu.wpi.first.math.trajectory.constraint.DifferentialDriveVoltageConstraint;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.units.measure.LinearAcceleration;
import edu.wpi.first.units.measure.LinearVelocity;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Filesystem;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.RamseteCommand;
import frc.robot.Constants.PathWeaverConstantsMargaret;
import frc.robot.Constants.PathWeaverConstantsSally;
import frc.robot.subsystems.Drivebase;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;

public class Trajectorygenerator {
  public static Command GetCommandForTrajectory(String fileToLoad, Drivebase drivebase) {
    final Distance TRACK_WIDTH_METERS = Meters.of(0.5588);
    final DifferentialDriveKinematics kDriveKinematics =
        new DifferentialDriveKinematics(TRACK_WIDTH_METERS);

    final LinearVelocity kMaxSpeed = MetersPerSecond.of(3);
    final LinearAcceleration kMaxAcceleration = MetersPerSecondPerSecond.of(3.0);

    final double kRamseteB = 2;
    final double kRamseteZeta = 0.7;

    DifferentialDriveVoltageConstraint autoVoltageConstraint;

    // Create a voltage constraint to ensure we don't accelerate too fast
    if (ConditionalConstants.SALLY) {
      autoVoltageConstraint = new DifferentialDriveVoltageConstraint(
          new SimpleMotorFeedforward(PathWeaverConstantsSally.kS, PathWeaverConstantsSally.kV,
              PathWeaverConstantsSally.kA),
          kDriveKinematics, 10);
    } else { // margeaert
      autoVoltageConstraint = new DifferentialDriveVoltageConstraint(
          new SimpleMotorFeedforward(PathWeaverConstantsMargaret.kS, PathWeaverConstantsMargaret.kV,
              PathWeaverConstantsMargaret.kA),
          kDriveKinematics, 10);
    }

    // Create config for trajectory
    TrajectoryConfig config = new TrajectoryConfig(kMaxSpeed, kMaxAcceleration)
                                  // Add kinematics to ensure max speed is actually obeyed
                                  .setKinematics(kDriveKinematics)
                                  // Apply the voltage constraint
                                  .addConstraint(autoVoltageConstraint);

    config.setReversed(true);

    // An example trajectory to follow. All units in meters.
    /*
    Trajectory trajectory =
        TrajectoryGenerator.generateTrajectory(
            // Start at the origin facing the +X direction
            new Pose2d(0, 0, new Rotation2d(0)),
            // Pass through these two interior waypoints, making an 's' curve path
            // TODO: Add the waypoints back in again.....
            List.of(),
            // End 3 meters straight ahead of where we started, facing forward
            new Pose2d(-3, 0, new Rotation2d(0)),
            // Pass config
            config);
    */

    String pathName = "output/" + fileToLoad + ".wpilib.json";
    Trajectory trajectory = new Trajectory();

    try {
      Path trajectoryPath = Filesystem.getDeployDirectory().toPath().resolve(pathName);
      trajectory = TrajectoryUtil.fromPathweaverJson(trajectoryPath);
    } catch (IOException ex) {
      DriverStation.reportError("Unable to open trajectory: " + pathName, ex.getStackTrace());
    }

    PrintTrajectoryStates(trajectory);

    RamseteCommand ramseteCommand;

    if (ConditionalConstants.SALLY) {
      ramseteCommand = new RamseteCommand(trajectory, drivebase::getPose,
          new RamseteController(kRamseteB, kRamseteZeta),
          new SimpleMotorFeedforward(PathWeaverConstantsSally.kS, PathWeaverConstantsSally.kV,
              PathWeaverConstantsSally.kA),
          kDriveKinematics, drivebase::getWheelSpeeds,
          new PIDController(PathWeaverConstantsSally.kP, 0, 0),
          new PIDController(PathWeaverConstantsSally.kP, 0, 0),
          // RamseteCommand passes volts to the callback
          drivebase::setVoltages, drivebase);
    } else { // margaret
      ramseteCommand = new RamseteCommand(trajectory, drivebase::getPose,
          new RamseteController(kRamseteB, kRamseteZeta),
          new SimpleMotorFeedforward(PathWeaverConstantsSally.kS, PathWeaverConstantsSally.kV,
              PathWeaverConstantsSally.kA),
          kDriveKinematics, drivebase::getWheelSpeeds,
          new PIDController(PathWeaverConstantsSally.kP, 0, 0),
          new PIDController(PathWeaverConstantsSally.kP, 0, 0),
          // RamseteCommand passes volts to the callback
          drivebase::setVoltages, drivebase);
    }

    // Reset odometry to the initial pose of the trajectory, run path following
    // command, then stop at the end.
    return Commands.runOnce(() -> drivebase.stop())
        .andThen(ramseteCommand)
        // Because Mr. Healy is professionally paranoid...
        .andThen(Commands.runOnce(() -> drivebase.stop()));
  }

  public static Pose2d GetTrajectoryInitialPose(String fileToLoad) {
    String pathName = "output/" + fileToLoad + ".wpilib.json";
    Trajectory trajectory = new Trajectory();

    try {
      Path trajectoryPath = Filesystem.getDeployDirectory().toPath().resolve(pathName);
      trajectory = TrajectoryUtil.fromPathweaverJson(trajectoryPath);
    } catch (IOException ex) {
      DriverStation.reportError("Unable to open trajectory: " + pathName, ex.getStackTrace());
    }

    return trajectory.getInitialPose();
  }

  public static Pose2d GetTrajectoryFinalPose(String fileToLoad) {
    String pathName = "output/" + fileToLoad + ".wpilib.json";
    Trajectory trajectory = new Trajectory();

    try {
      Path trajectoryPath = Filesystem.getDeployDirectory().toPath().resolve(pathName);
      trajectory = TrajectoryUtil.fromPathweaverJson(trajectoryPath);
    } catch (IOException ex) {
      DriverStation.reportError("Unable to open trajectory: " + pathName, ex.getStackTrace());
    }

    List<State> states = trajectory.getStates();
    return states.get(states.size() - 1).poseMeters; // last state
  }

  public static void PrintTrajectoryStates(Trajectory trajectory) {
    List<State> states = trajectory.getStates();

    for (State state : states) {
      System.out.println(state.toString());
    }
  }
}
